cmd/releasego/repo-release.go

```diff
@@ -0,0 +1,175 @@
+
+func (r repoReleaseCmd) Name() string {
+	return "repo-release"
+}
+
+func (r repoReleaseCmd) Summary() string {
+	return "Create a release on a GitHub repository."
+}
+
+func (r repoReleaseCmd) Description() string {
+	return `
+
+Using the GitHub API, create a release on the GitHub repository using a given tag name. Attach the
+given build asset JSON file and the artifacts it lists that are found in the specified directory.
+`
+}
+
+func (r repoReleaseCmd) Handle(p subcmd.ParseFunc) error {
+	tag := tagFlag()
+	repo := repoFlag()
+	pat := githubPATFlag()
+	buildAssetJSON := flag.String("build-asset-json", "", "[Required] The build asset JSON file to release.")
+	buildDir := flag.String("build-dir", "", "[Required] The directory containing build artifacts to attach.")
+
+	if err := p(); err != nil {
+		return err
+	}
+
+	if *tag == "" {
+		return fmt.Errorf("no tag specified")
+	}
+	owner, name, err := parseRepoFlag(*repo)
+	if err != nil {
+		return err
+	}
+	if *buildAssetJSON == "" {
+		return errors.New("no build asset json specified")
+	}
+	if *buildDir == "" {
+		return errors.New("no build dir specified")
+	}
+
+	var assets buildassets.BuildAssets
+	if err := buildmodel.ReadJSONFile(*buildAssetJSON, &assets); err != nil {
+		return err
+	}
+	uploadPaths := assetPaths(*buildDir, assets.GoSrcURL)
+	uploadPaths = append(uploadPaths, *buildAssetJSON)
+	log.Println("First, creating draft release. Then, attaching these files before marking release ready:")
+	for _, p := range uploadPaths {
+		log.Println(p)
+	}
+
+	ctx := context.Background()
+	client, err := githubClient(ctx, *pat)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Creating draft release %v...\n", *tag)
+
+	var release *github.RepositoryRelease
+	if err := retry(func() error {
+		release, _, err = client.Repositories.CreateRelease(ctx, owner, name, draftRelease(tag))
+		return err
+	}); err != nil {
+		return fmt.Errorf("unable to create draft release: %w", err)
+	}
+
+	// If we don't make it to the end of the func, clean up our partial-progress draft.
+	defer func() {
+		if release != nil {
+			log.Println("Cleaning up draft release.")
+			if err := retry(func() error {
+				_, err = client.Repositories.DeleteRelease(ctx, owner, name, *release.ID)
+				return err
+			}); err != nil {
+				// Log error and ignore. We're only trying to clean up.
+				log.Printf("Error trying to clean up draft release: %v", err)
+			}
+		}
+	}()
+
+	for _, p := range uploadPaths {
+		filename := filepath.Base(p)
+		log.Printf("Attaching (uploading) %#q\n", filename)
+		if err := retry(func() error {
+			file, err := os.Open(p)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+
+			_, _, err = client.Repositories.UploadReleaseAsset(
+				ctx,
+				owner, name,
+				*release.ID,
+				&github.UploadOptions{Name: filename},
+				file)
+			return err
+		}); err != nil {
+			return fmt.Errorf("failed to upload %#q to release %v: %w", p, *release.ID, err)
```

Comment:
If we fail to upload a file, we should fail the release. Does this happen here? Possibly it does, just my lack of go knowledge showing :)

Reply:
Ah, we're outside of the `retry` so this will return from `Handle` which will kick in the defered cleanup

Reply:
Yep, `err := retry(...)` gets the last inner error *if* it exceeds the retry count, then `if err := retry(...); err != nil {` continues here, and this return is in the function's scope.